### PR TITLE
bootutil: Fix wrong decryption attempt on invalid images

### PR DESCRIPTION
**Pull request by jamike**
With swap, and encryption active : virgin primary slots on flash
having 0xff erased pattern are detected as encrypted, and install
from secondary slot fails, when primary slot is fully erased

Signed-off-by: Michel Jaouen <michel.jaouen@st.com>

### CODE REVIEW

**Review by d3zd3z — APPROVED** [review]

**Review by d3zd3z — CHANGES_REQUESTED** [review]
I'm now beginning to think that this fix might not be in the right place. Really, the code should never be getting to a point where it is checking `IS_ENCRYPTED`, unless it has already decided that the header is valid. Have you found the path through the code that causes this, and I think there is either some earlier check missing, or we need to fix a check that isn't working right.

File: boot/bootutil/include/bootutil/image.h [hunk 1]
```diff
@@ -142,7 +142,7 @@ struct image_tlv {
     uint16_t it_len;    /* Data length (not including TLV header). */
 };
 
-#define IS_ENCRYPTED(hdr) ((hdr)->ih_flags & IMAGE_F_ENCRYPTED)
+#define IS_ENCRYPTED(hdr) (((hdr)->ih_flags & (IMAGE_F_ENCRYPTED)) && (hdr->ih_magic == IMAGE_MAGIC))
```
Comment (utzig) [inline boot/bootutil/include/bootutil/image.h:145]: I would suggest trying something like:

```
#define IS_ENCRYPTED(hdr) (((hdr)->ih_flags & IMAGE_F_ENCRYPTED) && ((hdr)->ih_flags != 0xffffffff))
```

Comment (d3zd3z) [reply boot/bootutil/include/bootutil/image.h:145]: This kind of implies that we won't ever use all of the flag bits. That might be a reasonable thing to do, especially if we decide to make a bit reserved, and indicate it must always be 0.

Additionally, not all flash devices use 0xffffffff as the erased value. Is there a particular reason to not check the magic? Arguably, that should be done first by the code using this, rather than every time the flags are checked, though.

Comment (jamike) [reply boot/bootutil/include/bootutil/image.h:145]: A possible fix is to clear ih_flags after having read it from flash . A  check of  hdr magic is added in function boot_read_image_header from swap_move.c or swap_scratch.c, if magic is not present all ih_flags are cleared.

Comment (jamike) [reply boot/bootutil/include/bootutil/image.h]: Updated patch fixes installation of encrypted file in secondary slot to an empty primary slot in SWAP_USING_MOVE  and SWAP_USING_SCRATCH

File: boot/bootutil/src/swap_move.c [hunk 2]
```diff
@@ -104,8 +104,9 @@ boot_read_image_header(struct boot_loader_state *state, int slot,
 
     /* We only know where the headers are located when bs is valid */
     if (bs != NULL && out_hdr->ih_magic != IMAGE_MAGIC) {
-        rc = -1;
-        goto done;
+        /*  image is not valid, unset all flags to avoid futher
+        *  unexpected behaviour */
```
Comment (gustavonihei) [inline boot/bootutil/src/swap_move.c:108]:
```suggestion
         *  unexpected behaviour */
```
nit: fixing code alignment

File: boot/bootutil/src/swap_move.c [hunk 3]
```diff
@@ -104,8 +104,9 @@ boot_read_image_header(struct boot_loader_state *state, int slot,
 
     /* We only know where the headers are located when bs is valid */
     if (bs != NULL && out_hdr->ih_magic != IMAGE_MAGIC) {
-        rc = -1;
-        goto done;
+        /*  image is not valid, unset all flags to avoid futher
```
Comment (gustavonihei) [inline boot/bootutil/src/swap_move.c:107]:
```suggestion
        /*  image is not valid, unset all flags to avoid further
```
Typo in `further`

File: boot/bootutil/src/swap_scratch.c [hunk 4]
```diff
@@ -73,6 +73,12 @@ boot_read_image_header(struct boot_loader_state *state, int slot,
         rc = BOOT_EFLASH;
     }
 
+    if (out_hdr->ih_magic != IMAGE_MAGIC) {
+       /*  image is not valid, unset all flags to avoid futher
```
Comment (gustavonihei) [inline boot/bootutil/src/swap_scratch.c:77]:
```suggestion
       /*  Image is not valid, unset all flags to avoid further
```
nit: Typo in `further` and capital initial on `Image`.

File: boot/bootutil/src/swap_move.c [hunk 5]
```diff
@@ -104,8 +104,9 @@ boot_read_image_header(struct boot_loader_state *state, int slot,
 
     /* We only know where the headers are located when bs is valid */
     if (bs != NULL && out_hdr->ih_magic != IMAGE_MAGIC) {
-        rc = -1;
```
Comment (nvlsianpu) [inline boot/bootutil/src/swap_move.c:107]: why -1 is not returned anymore?

File: boot/bootutil/src/loader.c [hunk 6]
```diff
@@ -107,7 +107,7 @@ boot_read_image_headers(struct boot_loader_state *state, bool require_all,
              *
              * Failure to read any headers is a fatal error.
              */
-            if (i > 0 && !require_all) {
+            if (rc < 0 && !require_all) {
```
Comment (gustavonihei) [inline boot/bootutil/src/loader.c:110]: Could you please explain why this change is required?
I understand the need to check for a negative `rc`, but considering the above comment, the check `i > 0` seems to be required.

File: boot/bootutil/src/swap_move.c [hunk 7]
```diff
@@ -104,7 +104,10 @@ boot_read_image_header(struct boot_loader_state *state, int slot,
...
+        /*  Image is not valid, unset all flags to avoid further
+         *  unexpected behaviour */
+        out_hdr->ih_flags = 0;
+        rc=-1;
```
Comment (gustavonihei) [inline boot/bootutil/src/swap_move.c:110]:
```suggestion
        rc = -1;
```

File: boot/bootutil/src/swap_move.c (hunk 5 above)
Comment (gustavonihei) [reply boot/bootutil/src/swap_move.c:107]: So you've added the `rc = -1` back to the Move implementation. Shouldn't Swap Scratch also return `-1` in this scenario?

File: boot/bootutil/src/loader.c (hunk 6 above)
Comment (jamike) [reply boot/bootutil/src/loader.c:110]: With i >0 an error is return if primary slot has no valid header, it prevents from parsing the secondary slot header, and installing an image on a secondary slot. Your remark is valid, and I have  another fix that I should update soon, there is more changes... Let's see automatic test result and then the review comment.
